### PR TITLE
perf: spawn jsr futures on executor

### PR DESCRIPTION
**Pull request by littledivy**
Adds an abstract trait for executing futures onto an async runtime. 

A default Tokio spawn executor is included behing the `tokio_executor` feature flag.

Closes https://github.com/denoland/deno_graph/issues/394

### CODE REVIEW

File: lib/lib.rs [hunk 1]
```diff
@@ -195,6 +195,22 @@ impl Resolver for JsResolver {
...
 }
 
+struct WasmExecutor;
+
+impl deno_graph::Executor for WasmExecutor {
+  fn execute(
+    &self,
+    future: futures::future::LocalBoxFuture<'static, ()>,
+  ) -> futures::future::LocalBoxFuture<'static, ()> {
+    // TODO(littledivy): bug in wasm_bindgen where Shared futures panic on spawn.
```
Comment (littledivy) [inline lib/lib.rs:205]: https://github.com/rustwasm/wasm-bindgen/issues/2562

**Comment by littledivy** [comment]
I can't test the perf in Deno because it is a few deno_graph versions behind it seems. Are there any other benchmarks I can run? @dsherret

**Comment by dsherret** [comment]
@littledivy it's up todate. Pull the latest main.

File: src/rt.rs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+/// An executor for futures.
+///
+/// This trait allows deno_graph to run background tasks on
+/// the async executor.
+pub trait Executor {
+  /// Spawns a future to run on this executor.
+  fn execute(&self, fut: BoxedFuture) -> BoxedFuture;
+}
+
+impl<'a> Default for &'a dyn Executor {
```
Comment (dsherret) [inline src/rt.rs:20]: Cool, I did not know it was possible to do this for traits.

**Review by dsherret — COMMENTED** [review]
Thanks so much for looking into this!

File: lib/lib.rs [hunk 3]
```diff
@@ -195,6 +195,22 @@ impl Resolver for JsResolver {
   }
 }
 
+struct WasmExecutor;
```
[the remark this answers is not in the document]
Comment (littledivy) [reply lib/lib.rs:198]: We could add it as an executor behind a feature flag similar to `tokio_executor`

[the remark this answers is not in the document]
Comment (littledivy) [reply lib/lib.rs]: Added it as `wasm_executor` feature flag.

**Comment by littledivy** [comment]
2.5x faster on incremental:

```
divy@divy:~/gh/denoland/jsr_benchmarks$ poop 'deno cache --no-check --no-config /home/divy/gh/denoland/jsr_benchmarks/dax_std/mod_jsr.ts' '/home/divy/gh/littledivy/deno/target/release/deno cache --no-check --no-config /home/divy/gh/denoland/jsr_benchmarks/dax_std/mod_jsr.ts'
Benchmark 1 (20 runs): deno cache --no-check --no-config /home/divy/gh/denoland/jsr_benchmarks/dax_std/mod_jsr.ts
  measurement          mean ± σ            min … max           outliers         delta
  wall_time           254ms ±  333ms     177ms … 1.67s           1 ( 5%)        0%
  peak_rss           62.9MB ± 1.38MB    62.1MB … 68.6MB          1 ( 5%)        0%
  cpu_cycles          652M  ±  169M      606M  … 1.37G           1 ( 5%)        0%
  instructions       1.01G  ±  239M      961M  … 2.03G           1 ( 5%)        0%
  cache_references   45.0M  ± 11.4M     42.0M  … 93.5M           1 ( 5%)        0%
  cache_misses       9.77M  ± 4.34M     8.61M  … 28.2M           1 ( 5%)        0%
  branch_misses      3.44M  ± 1.35M     3.09M  … 9.18M           1 ( 5%)        0%
Benchmark 2 (50 runs): /home/divy/gh/littledivy/deno/target/release/deno cache --no-check --no-config /home/divy/gh/denoland/jsr_benchmarks/dax_std/mod_jsr.ts
  measurement          mean ± σ            min … max           outliers         delta
  wall_time           100ms ±  217ms    68.1ms … 1.61s           2 ( 4%)        ⚡- 60.5% ± 53.2%
  peak_rss           39.9MB ± 4.06MB    39.0MB … 68.0MB          1 ( 2%)        ⚡- 36.5% ±  3.0%
  cpu_cycles          200M  ±  136M      177M  … 1.14G           1 ( 2%)        ⚡- 69.4% ± 11.8%
  instructions        310M  ±  208M      281M  … 1.75G           1 ( 2%)        ⚡- 69.4% ± 11.3%
  cache_references   11.9M  ± 9.13M     10.5M  … 75.2M           2 ( 4%)        ⚡- 73.5% ± 11.6%
  cache_misses       3.77M  ± 2.85M     3.25M  … 23.5M           1 ( 2%)        ⚡- 61.4% ± 18.1%
  branch_misses      1.02M  ± 1.05M      844K  … 8.27M           1 ( 2%)        ⚡- 70.4% ± 17.5%
```


5+ seconds faster with no cache (network is involved):
```
time /home/divy/gh/littledivy/deno/target/release/deno cache --no-check --no-config -r /home/divy/gh/denoland/jsr_benchmarks/dax_std/mod_jsr.ts

real	0m15.635s
user	0m0.067s
sys	0m0.034s
```

```
time deno cache --no-check --no-config -r /home/divy/gh/denoland/jsr_benchmarks/dax_std/mod_jsr.ts

real	0m20.879s
user	0m0.301s
sys	0m0.188s
```

File: lib/Cargo.toml [hunk 4]
```diff
@@ -16,7 +16,7 @@ crate-type = ["cdylib"]
 [dependencies]
 anyhow = "1.0.43"
 console_error_panic_hook = "0.1.7"
-deno_graph = { path = "../" }
+deno_graph = { path = "../", default-features = false, features = ["fast_check", "wasm_executor"] }
```
Comment (dsherret) [inline lib/Cargo.toml:19]:
```suggestion
deno_graph = { path = "../", default-features = false, features = ["wasm_executor"] }
```

**Review by dsherret — APPROVED** [review]
LGTM!

**Pull request merged**